src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -1226,19 +1005,73 @@ public int hashCode() {
         }
     }
 
-    private void logNodeMessage(FlowNode node) {
-        WorkflowConsoleLogger wfLogger = new WorkflowConsoleLogger(getListener());
-        String prefix = getBranchName(node);
-        if (prefix != null) {
-            wfLogger.log(String.format("[%s] %s", prefix, node.getDisplayFunctionName()));
-        } else {
-            wfLogger.log(node.getDisplayFunctionName());
+    /**
+     * Prints nodes as they appear (including block start and end nodes).
+     */
+    private final class NodePrintListener implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            NewNodeConsoleNote.print(node, getListener());
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override public AnnotatedLargeText getLogText() {
+        return LogStorage.of(asFlowExecutionOwner()).overallLog(this, !isLogUpdated());
+    }
+
+    // TODO log-related overrides pending JEP-207:
+
+    @Override public InputStream getLogInputStream() throws IOException {
+        // Inefficient but probably rarely used anyway.
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        getLogText().writeRawLogTo(0, baos);
+        return new ByteArrayInputStream(baos.toByteArray());
+    }
+
+    @Override public Reader getLogReader() throws IOException {
+        return getLogText().readAll();
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override public String getLog() throws IOException {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        getLogText().writeRawLogTo(0, baos);
+        return baos.toString("UTF-8");
+    }
+
+    @Override public List<String> getLog(int maxLines) throws IOException {
+        int lineCount = 0;
+        List<String> logLines = new LinkedList<>();
+        if (maxLines == 0) {
+            return logLines;
+        }
+        try (BufferedReader reader = new BufferedReader(getLogReader())) {
+            for (String line = reader.readLine(); line != null; line = reader.readLine()) {
+                logLines.add(line);
+                ++lineCount;
+                if (lineCount > maxLines) {
+                    logLines.remove(0);
+                }
+            }
+        }
+        if (lineCount > maxLines) {
+            logLines.set(0, "[...truncated " + (lineCount - (maxLines - 1)) + " lines...]");
```

Comment:
🐛 You just disappeared the first line of log output.  I think you mean to *insert* there.  And the method should note that it actually returns maxLines+1 lines due to the prefix.

Reply:
Ditto, copied from core.

Reply:
@jglick A bug that exists in TWO places is even worse than a bug existing in one, not better...

Reply:
Or is there a different argument you're trying to make here?

Reply:
This code is _temporarily_ copied from core pending JEP-207 or some other patch to core to delegate to `getLogText`. As soon as core does proper delegation, this override can be removed, so there is no reason to waste time fixing (supposed) buglets in code which has existed for years and is not really part of this PR. Anyway this method is never used in production code, only test output.

Reply:
The Javadoc for `Run#maxLines` says:

> The maximum number of log lines to return.  If the log is bigger than this, only the most recent lines are returned.

Given that documentation, it doesn't seem like returning `maxLines-1` of log content and one line explaining that things were truncated is a problem. Besides, in `Run#getLog` there is [a comment](https://github.com/jenkinsci/jenkins/blob/c4acc81405068fd969272fdb56c19da161a7015b/core/src/main/java/hudson/model/Run.java#L2079) explaining that `List#set` is used intentionally to avoid returning more lines:

> // If the log has been truncated, include that information.
> // Use set (replaces the first element) rather than add so that
> // the list doesn't grow beyond the specified maximum number of lines.